src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -105,6 +119,25 @@ void BaseCouplingScheme::receiveData(const m2n::PtrM2N &m2n, const DataMap &rece
     // Data is only received on ranks with size>0, which is checked in the derived class implementation
     m2n->receive(pair.second->values(), pair.second->getMeshID(), pair.second->getDimensions());
 
+    if (pair.second->hasGradient())
+    {
+      Eigen::VectorXd gradVals0 = pair.second->gradientValues().row(0);
+      m2n->receive(gradVals0, pair.second->getMeshID(), pair.second->getDimensions(), true, 0);
+
+      if (pair.second->gradientValues().rows() > 1)
+      {
+        Eigen::VectorXd gradVals1 = pair.second->gradientValues().row(1);
+        m2n->receive(gradVals1, pair.second->getMeshID(), pair.second->getDimensions(), true, 1);
+      }
+
+      if (pair.second->gradientValues().rows() > 2)
+      {
+        Eigen::VectorXd gradVals2 = pair.second->gradientValues().row(2);
+        m2n->receive(gradVals2, pair.second->getMeshID(), pair.second->getDimensions(), true, 2);
+      }
```

Comment:
This won't work since you copy the data when you call `Eigen::VectorXd gradVals0 = pair.second->gradientValues().row(0);`. In order to get a reference, you need to call `Eigen::VectorXd &gradVals0` with an ampersand (lvalue reference). This is also the return type of `pair.second->values()`. 
Although this doesn't change the values when sending data (you send the same data) please also change this in the send routine, because otherwise you copy just unnecessary data around.

This should fix your problem.